prometheus/exporter/watchdog.py

```diff
@@ -329,43 +369,58 @@ def main(argv):
+        k8s_gauge = gen_k8s_component_gauge()
 
+        try:
             # 1. check service level status
-            podsStatus = requests.get("{}/api/v1/namespaces/default/pods/".format(address)).json()
-            pods_metrics = parse_pods_status(podsStatus)
-            if pods_metrics is not None:
-                metrics.extend(pods_metrics)
+            podsStatus = request_with_histogram(list_pods_url, list_pods_histogram)
+            process_pods_status(pai_pod_gauge, pai_container_gauge, podsStatus)
 
             # 2. check nodes level status
-            nodesStatus = requests.get("{}/api/v1/nodes/".format(address)).json()
-            nodes_metrics = parse_nodes_status(nodesStatus)
-            if nodes_metrics is not None:
-                metrics.extend(nodes_metrics)
+            nodesStatus = request_with_histogram(list_nodes_url, list_nodes_histogram)
+            process_nodes_status(pai_node_gauge, nodesStatus)
 
             # 3. check docker deamon status
-            docker_daemon_metrics = collect_docker_daemon_status(hosts)
-            if docker_daemon_metrics is not None:
-                metrics.extend(docker_daemon_metrics)
+            collect_docker_daemon_status(docker_daemon_gauge, hosts)
 
             # 4. check k8s level status
-            k8s_metrics = collect_k8s_componentStaus(api_server_ip, api_server_port, nodesStatus)
-            if k8s_metrics is not None:
-                metrics.extend(k8s_metrics)
-
-            # 5. log and export
-            log_and_export_metrics(logDir + "/watchdog.prom", metrics)
+            collect_k8s_componentStaus(k8s_gauge, api_server_ip, api_server_port, nodesStatus)
         except Exception as e:
+            error_counter.labels(type="unknown").inc()
             logger.exception("watchdog failed in one iteration")
 
-            # do not lost metrics due to exception
-            log_and_export_metrics(logDir + "/watchdog.prom", metrics)
+        atomic_ref.get_and_set([pai_pod_gauge, pai_container_gauge, pai_node_gauge,
+            docker_daemon_gauge, k8s_gauge])
 
-        time.sleep(timeSleep)
+        time.sleep(float(args.interval))
 
-# python watch_dog.py /datastorage/prometheus /usr/local/cluster-configuration.yaml 3
-# requires env K8S_API_SERVER_URI set to correct value, eg. http://10.151.40.133:8080
+# python watchdog.py http://10.151.40.133:8080
 if __name__ == "__main__":
-    main(sys.argv[1:])
+    parser = argparse.ArgumentParser()
+    parser.add_argument("k8s_api", help="kubernetes api uri eg. http://10.151.40.133:8080")
+    parser.add_argument("--log", "-l", help="log dir to store log", default="/datastorage/prometheus")
+    parser.add_argument("--interval", "-i", help="interval between two collection", default="30")
+    parser.add_argument("--port", "-p", help="port to expose metrics", default="9101")
+    parser.add_argument("--hosts", "-m", help="yaml file path contains host info", default="/etc/watchdog/config.yml")
```

Comment:
Currently, how to generate the hosts' yaml file

Reply:
by using `pai-management/bootstrap/prometheus/watchdog-configmap.yaml.template`